cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -237,13 +247,58 @@ private void processTetheringControlResponse(String message, PeerInfo peerInfo)
-    try {
-      messagePublisher.publish(topic.getNamespace(), topic.getTopic(), StandardCharsets.UTF_8, GSON.toJson(message));
-    } catch (TopicNotFoundException | IOException e) {
-      throw new IOException(String.format("Failed to publish message to topic %s", topic.getTopic()), e);
+  private void publishRunPipeline(String runPayload, String peerName) throws IOException {
+    TetheringLaunchMessage message = GSON.fromJson(runPayload, TetheringLaunchMessage.class);
+
+    CConfiguration cConfCopy = CConfiguration.copy(cConf);
+    for (Map.Entry<String, String> cConfEntry : message.getCConfEntries().entrySet()) {
+      cConfCopy.set(cConfEntry.getKey(), cConfEntry.getValue());
+    }
+
+    Map<String, String> files = new HashMap<>();
+    for (Map.Entry<String, byte[]> file : message.getFiles().entrySet()) {
+      try (GZIPInputStream inputStream = new GZIPInputStream(new ByteArrayInputStream(file.getValue()))) {
+        files.put(file.getKey(), IOUtils.toString(inputStream));
+      }
+    }
+
+    ApplicationSpecification appSpec = GSON.fromJson(files.get(DistributedProgramRunner.APP_SPEC_FILE_NAME),
+                                                     ApplicationSpecification.class);
+    ProgramOptions programOpts = GSON.fromJson(files.get(DistributedProgramRunner.PROGRAM_OPTIONS_FILE_NAME),
+                                               ProgramOptions.class);
+    ProgramId programId = new ProgramId(message.getNamespace(), programOpts.getProgramId().getApplication(),
+                                        programOpts.getProgramId().getType(), programOpts.getProgramId().getProgram());
+    Map<String, String> systemArgs = new HashMap<>(programOpts.getArguments().asMap());
+    systemArgs.put(ProgramOptionConstants.PEER_NAME, peerName);
+    ProgramOptions updatedOpts = new SimpleProgramOptions(programId, new BasicArguments(systemArgs),
+                                                          programOpts.getUserArguments());
+    ProgramRunId programRunId = programId.run(programOpts.getArguments().getOption(ProgramOptionConstants.RUN_ID));
+    ProgramDescriptor programDescriptor = new ProgramDescriptor(programId, appSpec);
+
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    File runDir = new File(tmpDir, programRunId.getRun());
+    runDir.mkdirs();
+    Files.write(files.get(DistributedProgramRunner.LOGBACK_FILE_NAME),
+                new File(runDir, DistributedProgramRunner.LOGBACK_FILE_NAME), StandardCharsets.UTF_8);
+    try (Writer writer = Files.newWriter(new File(runDir, "cdap-tether.xml"), StandardCharsets.UTF_8)) {
+      cConfCopy.writeXml(writer);
     }
+
+    programStateWriter.start(programRunId, updatedOpts, null, programDescriptor);
+    LOG.debug("Updated program run {} state to started", programRunId.getRun());
```

Comment:
state update happens later by the class that subscribes to the topic. It's more accurate to say something like  `Published program start message for run {}`